### PR TITLE
BUG 1686943: asset/store: include all assets to preserve when doing a fetch

### PR DESCRIPTION
**Pull request by staebler**
Pass in to the Fetch function all of the assets that should not be consumed when purging.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1686943

### CODE REVIEW

**Comment by staebler** [comment]
```console
$ cat <<EOF > install-config.yaml
apiVersion: v1beta4
baseDomain: devcluster.openshift.com
metadata:
  name: mstaeble
platform:
  aws:
    region: us-east-2
pullSecret: redacted
sshKey: redacted

$ bin/openshift-install create manifests
INFO Consuming "Install Config" from target directory

$ tree
.
├── manifests
│   ├── 04-openshift-machine-config-operator.yaml
│   ├── cluster-config.yaml
│   ├── cluster-dns-02-config.yml
│   ├── cluster-infrastructure-02-config.yml
│   ├── cluster-ingress-02-config.yml
│   ├── cluster-network-01-crd.yml
│   ├── cluster-network-02-config.yml
│   ├── cvo-overrides.yaml
│   ├── etcd-service.yaml
│   ├── host-etcd-service-endpoints.yaml
│   ├── host-etcd-service.yaml
│   ├── kube-cloud-config.yaml
│   ├── kube-system-configmap-etcd-ca-bundle.yaml
│   ├── kube-system-configmap-etcd-serving-ca.yaml
│   ├── kube-system-configmap-root-ca.yaml
│   ├── kube-system-secret-etcd-client-ca-deprecated.yaml
│   ├── kube-system-secret-etcd-client.yaml
│   ├── kube-system-secret-etcd-signer-client.yaml
│   ├── kube-system-secret-etcd-signer.yaml
│   ├── machine-config-server-tls-secret.yaml
│   ├── openshift-config-configmap-etcd-metric-serving-ca.yaml
│   ├── openshift-config-secret-etcd-metric-client.yaml
│   └── pull.json
└── openshift
    ├── 99_binding-discovery.yaml
    ├── 99_cloud-creds-secret.yaml
    ├── 99_kubeadmin-password-secret.yaml
    ├── 99_openshift-cluster-api_master-machines-0.yaml
    ├── 99_openshift-cluster-api_master-machines-1.yaml
    ├── 99_openshift-cluster-api_master-machines-2.yaml
    ├── 99_openshift-cluster-api_master-user-data-secret.yaml
    ├── 99_openshift-cluster-api_worker-machineset-0.yaml
    ├── 99_openshift-cluster-api_worker-machineset-1.yaml
    ├── 99_openshift-cluster-api_worker-machineset-2.yaml
    ├── 99_openshift-cluster-api_worker-user-data-secret.yaml
    ├── 99_openshift-machineconfig_master.yaml
    ├── 99_openshift-machineconfig_worker.yaml
    └── 99_role-cloud-creds-secret-reader.yaml

2 directories, 37 files

$ bin/openshift-install create manifests

$ tree
.
├── manifests
│   ├── 04-openshift-machine-config-operator.yaml
│   ├── cluster-config.yaml
│   ├── cluster-dns-02-config.yml
│   ├── cluster-infrastructure-02-config.yml
│   ├── cluster-ingress-02-config.yml
│   ├── cluster-network-01-crd.yml
│   ├── cluster-network-02-config.yml
│   ├── cvo-overrides.yaml
│   ├── etcd-service.yaml
│   ├── host-etcd-service-endpoints.yaml
│   ├── host-etcd-service.yaml
│   ├── kube-cloud-config.yaml
│   ├── kube-system-configmap-etcd-ca-bundle.yaml
│   ├── kube-system-configmap-etcd-serving-ca.yaml
│   ├── kube-system-configmap-root-ca.yaml
│   ├── kube-system-secret-etcd-client-ca-deprecated.yaml
│   ├── kube-system-secret-etcd-client.yaml
│   ├── kube-system-secret-etcd-signer-client.yaml
│   ├── kube-system-secret-etcd-signer.yaml
│   ├── machine-config-server-tls-secret.yaml
│   ├── openshift-config-configmap-etcd-metric-serving-ca.yaml
│   ├── openshift-config-secret-etcd-metric-client.yaml
│   └── pull.json
└── openshift
    ├── 99_binding-discovery.yaml
    ├── 99_cloud-creds-secret.yaml
    ├── 99_kubeadmin-password-secret.yaml
    ├── 99_openshift-cluster-api_master-machines-0.yaml
    ├── 99_openshift-cluster-api_master-machines-1.yaml
    ├── 99_openshift-cluster-api_master-machines-2.yaml
    ├── 99_openshift-cluster-api_master-user-data-secret.yaml
    ├── 99_openshift-cluster-api_worker-machineset-0.yaml
    ├── 99_openshift-cluster-api_worker-machineset-1.yaml
    ├── 99_openshift-cluster-api_worker-machineset-2.yaml
    ├── 99_openshift-cluster-api_worker-user-data-secret.yaml
    ├── 99_openshift-machineconfig_master.yaml
    ├── 99_openshift-machineconfig_worker.yaml
    └── 99_role-cloud-creds-secret-reader.yaml

2 directories, 37 files

```

File: pkg/asset/store/store.go [hunk 1]
```diff
@@ -335,13 +336,16 @@ func (s *storeImpl) load(a asset.Asset, indent string) (*assetState, error) {
...
+STORE_ASSETS_LOOP:
 	for _, assetState := range s.assets {
 		if !assetState.presentOnDisk {
 			continue
 		}
-		if reflect.TypeOf(assetState.asset) == reflect.TypeOf(excluded) {
-			continue
+		for _, a := range excluded {
+			if reflect.TypeOf(assetState.asset) == reflect.TypeOf(a) {
+				continue STORE_ASSETS_LOOP
```
Comment (abhinavdahiya) [inline pkg/asset/store/store.go]: Hmm.. can we avoid the jump with map or internal func ?

Comment (staebler) [reply pkg/asset/store/store.go]: Fixed with c9ccee9be...d3d8cade3.

File: pkg/asset/store.go [hunk 2]
```diff
@@ -3,8 +3,9 @@ package asset
 // Store is a store for the states of assets.
 type Store interface {
 	// Fetch retrieves the state of the given asset, generating it and its
-	// dependencies if necessary.
-	Fetch(Asset) error
+	// dependencies if necessary. When purging consumed assets, none of the
+	// assets in assetsToPreserve will be purged.
+	Fetch(assetToFetch Asset, assetsToPreserve ...WritableAsset) error
```
Comment (abhinavdahiya) [inline pkg/asset/store.go]: what was the reason you dropped https://github.com/openshift/installer/pull/1531/files#diff-48efe5e4ccec738eec350b7816cc6ce2R7 in favor of explicit.

do you see that we might want to preserver only part of targets?

Comment (staebler) [reply pkg/asset/store.go]: I am trying to minimize the changes to reduce the contention. The approach here preserves the order in which operations are performed. For each asset in order, there is a fetch of the asset, a save of the state file, a purge of consumed assets, and then a write of the asset files to disk. If all of the assets are passed to `Fetch` as one slice, then none of the assets files would be written to disk until after all of the assets have been fetched.

**Pull request merged**